File: problems/HackerRank/2521_exercises/linkedList/intermediate/rotateRight.c
struct ListNode *rotateRight(struct ListNode *head, int k)
{
	if (head == NULL)
	{
		return head;
	}

	struct ListNode *tail = head;
	int length = 1;

	// count the length of the list
	while (tail->next != NULL)
	{
		tail = tail->next;
		length = length + 1;
	}

	k = k % length;
	if (k == 0)
		return head;

	struct ListNode *current = head;
	// find the pivet
	for (int i = 0; i < length - k - 1; i++)
	{
		current = current->next;
	}

	struct ListNode *newHead = current->next;
	current->next = NULL;
	tail->next = head;

	return newHead;
}

File: problems/LeetCode/TopInterview150/rotateRight.ts
function rotateRight(head: ListNode | null, k: number): ListNode | null {
  if (!head) return head;

  // get the length of the list
  let len = 1;
  let tail = head;

  while (tail.next) {
    len += 1;
    tail = tail.next;
  }

  k = k % len;
  if (k === 0) return head;

  // get the node before the new head
  let cur = head;
  for (let i = 0; i < len - k - 1; i++) {
    cur = cur.next;
  }

  let newHead = cur.next;
  cur.next = null;
  tail.next = head;

  return newHead;
}

// Time complexity: O(n)
// Space complexity: O(1)
